dmoj/cptbox/utils.py

```diff
@@ -1,11 +1,85 @@
+import errno
 import io
+import mmap
+import os
+from tempfile import NamedTemporaryFile
+from typing import Optional
 
 from dmoj.cptbox._cptbox import memory_fd_create, memory_fd_seal
 
 
-class MemoryIO(io.FileIO):
-    def __init__(self) -> None:
-        super().__init__(memory_fd_create(), 'r+')
+def _make_fd_readonly(fd):
+    new_fd = os.open(f'/proc/self/fd/{fd}', os.O_RDONLY)
+    try:
+        os.dup2(new_fd, fd)
+    finally:
+        os.close(new_fd)
+
+
+class MmapableIO(io.FileIO):
+    def __init__(self, fd, *, prefill: Optional[bytes] = None, seal=False) -> None:
+        super().__init__(fd, 'r+')
+
+        if prefill:
+            self.write(prefill)
+        if seal:
+            self.seal()
+
+    def to_bytes(self) -> bytes:
+        try:
+            with mmap.mmap(self.fileno(), 0, access=mmap.ACCESS_READ) as f:
```

Comment:
How often do we expect this will be called? Should we `madvise(..., MADV_SEQUENTIAL)` here?

Reply:
Not very often, it's mostly for compatibility with old checkers etc.

Reply:
Sounds like "very often" to me, but happy to punt on this. I worry we'll hit issues with gigabyte-sized generator inputs that also have checkers, since this doubles the memory requirement.

Reply:
It's less of a problem than it looks. In the standard grader, we pass this magic to checkers: `judge_input=LazyBytes(case.input_data)`. We only pay for this if the checker actually reads `judge_input`.